src/en/kmanga/src/eu/kanade/tachiyomi/extension/en/kmanga/KManga.kt

```diff
@@ -0,0 +1,369 @@
+            .addQueryParameter("ranking_id", "12")
+            .addQueryParameter("offset", "0")
+            .addQueryParameter("limit", "999") // 562 entries atm
+            .build()
+
+        val (birthday, expires) = getBirthdayCookie(url)
+
+        val queryParams = url.queryParameterNames.associateWith { url.queryParameter(it)!! }
+        val hash = generateHash(queryParams, birthday, expires)
+
+        val newHeaders = headersBuilder()
+            .add("x-kmanga-hash", hash)
+            .build()
+
+        return GET(url, newHeaders)
+    }
+
+    override fun popularMangaParse(response: Response): MangasPage {
+        val rankingResult = json.decodeFromString<RankingApiResponse>(response.body.string())
+        val titleIds = rankingResult.ranking_title_list.map { it.id.toString() }
+
+        if (titleIds.isEmpty()) {
+            return MangasPage(emptyList(), false)
+        }
+
+        val detailsUrl = "$apiUrl/title/list".toHttpUrl().newBuilder()
+            .addQueryParameter("title_id_list", titleIds.joinToString(","))
+            .build()
+
+        val (birthday, expires) = getBirthdayCookie(detailsUrl)
+        val queryParams = detailsUrl.queryParameterNames.associateWith { detailsUrl.queryParameter(it)!! }
+        val hash = generateHash(queryParams, birthday, expires)
+
+        val detailsHeaders = headersBuilder()
+            .add("x-kmanga-hash", hash)
+            .build()
+
+        val detailsRequest = GET(detailsUrl, detailsHeaders)
+        val detailsResponse = client.newCall(detailsRequest).execute()
+
+        if (!detailsResponse.isSuccessful) {
+            throw Exception("Failed to fetch title details: ${detailsResponse.code} - ${detailsResponse.body.string()}")
+        }
+
+        val detailsResult = json.decodeFromString<TitleListResponse>(detailsResponse.body.string())
+        val mangas = detailsResult.title_list.map { manga ->
+            SManga.create().apply {
+                url = "/title/${manga.title_id}"
+                title = manga.title_name
+                thumbnail_url = manga.thumbnail_rect_image_url ?: manga.banner_image_url
+            }
+        }
+        return MangasPage(mangas, false)
+    }
+
+    // Search
+    override fun fetchSearchManga(page: Int, query: String, filters: FilterList): Observable<MangasPage> {
+        val prefix = "id:"
+
+        return if (query.startsWith(prefix)) {
+            val titleId = query.removePrefix(prefix)
+            fetchMangaDetails(
+                SManga.create().apply { url = "/title/$titleId" },
+            ).map { manga ->
+                MangasPage(listOf(manga.apply { url = "/title/$titleId" }), false)
+            }
+        } else {
+            super.fetchSearchManga(page, query, filters)
+        }
+    }
+
+    override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request {
+        val genreFilter = filters.filterIsInstance<GenreFilter>().first()
```

Comment:
```suggestion
        val genreFilter = filters.firstInstance<GenreFilter>()
```
import from `keiyoushi.utils`